File: android/core/engine/src/main/kotlin/com/cardcopilot/engine/engine/CreditCheckoutAdvisor.kt
package com.cardcopilot.engine.engine

import com.cardcopilot.engine.models.CardProduct
import com.cardcopilot.engine.models.CardState
import com.cardcopilot.engine.models.CardCredit
import com.cardcopilot.engine.models.CreditRedemptionMethod
import com.cardcopilot.engine.models.Money
import com.cardcopilot.engine.models.OwnerState
import com.cardcopilot.engine.models.PurchaseContext
import com.cardcopilot.engine.models.ReportingCurrency
import com.cardcopilot.engine.models.SourceType
import kotlinx.serialization.Serializable

@Serializable
data class CheckoutCreditMatch(
    val cardId: String,
    val creditId: String,
    val label: String,
    val valueCad: Double,
    val redemptionMethod: CreditRedemptionMethod
)

/** Fail-closed POS credit matching. Only one credit is counted until stacking is contractual. */
object CreditCheckoutAdvisor {
    fun bestMatch(card: CardProduct, purchase: PurchaseContext,
                  ownerState: OwnerState, asOf: String): CheckoutCreditMatch? {
        val cardState = ownerState.cardStates[card.cardId] ?: CardState()
        return bestMatch(card.cardId, card.credits ?: emptyList(), cardState, purchase, asOf)
    }

    fun bestMatch(cardId: String, credits: List<CardCredit>, cardState: CardState,
                  purchase: PurchaseContext, asOf: String): CheckoutCreditMatch? {
        val canonical = purchase.canonicalized()
        return credits.mapNotNull { credit ->
            if (credit.sourceType != SourceType.ISSUER_CONFIRMED) return@mapNotNull null
            val method = credit.redemptionMethod ?: return@mapNotNull null
            val predicate = credit.purchasePredicate ?: return@mapNotNull null
            if (!RuleMatcher.matches(predicate, canonical, cardState)) return@mapNotNull null
            val opportunity = CreditAdvisor.opportunity(cardId, credit, cardState, asOf)
                ?: return@mapNotNull null
            if (opportunity.status != CreditOpportunityStatus.AVAILABLE) return@mapNotNull null

            val purchaseCad = maxOf(0.0, canonical.amountCad)
            val minimumCad = credit.minimumTransaction?.let(ReportingCurrency::toReporting)
            if (minimumCad != null && purchaseCad + 0.000001 < minimumCad) return@mapNotNull null
            val remainingCad = ReportingCurrency.toReporting(
                Money(opportunity.remainingAmount, credit.value.currency)
            )
            val valueCad = if (credit.allowsPartialUse == true) {
                minOf(remainingCad, purchaseCad)
            } else {
                if (purchaseCad + 0.000001 < remainingCad) return@mapNotNull null
                remainingCad
            }
            if (valueCad <= 0.000001) return@mapNotNull null
            CheckoutCreditMatch(cardId, credit.creditId, credit.label, valueCad, method)
        }.maxWithOrNull(compareBy<CheckoutCreditMatch> { it.valueCad }.thenByDescending { it.creditId })
    }
}

File: Engine/Sources/CardCopilotEngine/Engine/CreditCheckoutAdvisor.swift
import Foundation

public struct CheckoutCreditMatch: Equatable, Sendable {
    public let cardId: String
    public let creditId: String
    public let label: String
    /// Defensible value of this credit on this purchase in PickMe's CAD reporting currency.
    public let valueCad: Double
    public let redemptionMethod: CreditRedemptionMethod
}

/// Finds an issuer-confirmed, enrolled, unused credit that this exact purchase can recover.
/// Missing merchant/category eligibility fails closed. At most one credit is counted because the
/// contract does not yet prove that two credits stack on the same transaction.
public enum CreditCheckoutAdvisor {
    public static func bestMatch(card: CardProduct, purchase: PurchaseContext,
                                 ownerState: OwnerState, asOf: String) -> CheckoutCreditMatch? {
        let cardState = ownerState.cardStates[card.cardId] ?? CardState()
        return bestMatch(cardId: card.cardId, credits: card.credits ?? [], cardState: cardState,
                         purchase: purchase, asOf: asOf)
    }

    public static func bestMatch(cardId: String, credits: [CardCredit], cardState: CardState,
                                 purchase: PurchaseContext, asOf: String) -> CheckoutCreditMatch? {
        let purchase = purchase.canonicalized()
        return credits.compactMap { credit -> CheckoutCreditMatch? in
            guard credit.sourceType == .issuerConfirmed,
                  let method = credit.redemptionMethod,
                  let predicate = credit.purchasePredicate,
                  RuleMatcher.matches(predicate, purchase: purchase, state: cardState),
                  let opportunity = CreditAdvisor.opportunity(cardId: cardId, credit: credit,
                                                              cardState: cardState, asOf: asOf),
                  opportunity.status == .available else { return nil }

            let purchaseCad = max(0, purchase.amountCad)
            if let minimum = credit.minimumTransaction,
               purchaseCad + 0.000_001 < ReportingCurrency.toReporting(minimum) { return nil }

            let remainingCad = ReportingCurrency.toReporting(
                Money(amount: opportunity.remainingAmount, currency: credit.value.currency)
            )
            let valueCad: Double
            if credit.allowsPartialUse == true {
                valueCad = min(remainingCad, purchaseCad)
            } else {
                guard purchaseCad + 0.000_001 >= remainingCad else { return nil }
                valueCad = remainingCad
            }
            guard valueCad > 0.000_001 else { return nil }
            return CheckoutCreditMatch(cardId: cardId, creditId: credit.creditId,
                                       label: credit.label, valueCad: valueCad,
                                       redemptionMethod: method)
        }
        .max { ($0.valueCad, $1.creditId) < ($1.valueCad, $0.creditId) }
    }
}
